presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -211,7 +253,62 @@ public static JoinNode createJoinAccordingToPartitioning(JoinGraphNode joinGraph
                     joinFilters.isEmpty() ? Optional.empty() : Optional.of(ExpressionUtils.and(joinFilters)),
                     Optional.empty(),
                     Optional.empty(),
-                    Optional.empty());
+                    Optional.empty()));
+        }
+
+        private boolean introducesCrossJoin(List<PlanNode> leftSources, List<PlanNode> rightSources)
+        {
+            List<PlanNode> allSources = ImmutableList.<PlanNode>builder()
+                    .addAll(leftSources)
+                    .addAll(rightSources)
+                    .build();
+            Set<PlanNode> leftSet = ImmutableSet.copyOf(leftSources);
+            Set<PlanNode> rightSet = ImmutableSet.copyOf(rightSources);
+            Set<Set<PlanNode>> connectedNodeSets = getConnectedNodeSets(allSources);
+
+            for (Set<PlanNode> connectedSet : connectedNodeSets) {
```

Comment:
I don't follow what you mean.  It may not be possible for the left and right set to be connected (especially if you have >1 cross join in the query).  To know that you'd need to know about the connectedness of the original graph and whether this makes some part that had been connected disconnected.  but see my comment at https://github.com/Teradata/presto/pull/559#discussion_r120720429 and let me know what you think of the new proposal (you can check out the commit called New cross join strategy too for the corresponding code).